### ISSUE TITLE
Wrong line numbers produces in rustc_span::SourceFile::lookup_file_pos

### ISSUE THREAD

**Issue by klensy** [rust-lang/rust#84127]
Function comments noted, that they returns 1-based line number, but can actually return 0-based, #77080 @richkadel
https://github.com/rust-lang/rust/blob/c18c0ad2bc5988ca7953459e5a35ece8e69e35e7/compiler/rustc_span/src/lib.rs#L1515-L1573

Perhaps this can be prevented via using different types for 0- and 1- based line numbers.

**Comment by richkadel** [comment]
You're absolutely right that this doc comment is incomplete, though I would not say it "0-based" (valid values are always 1-based). 

But I should correct the comment. The source uses `lookup_line()`, which can return `None`. Adapting the description from that function, we should add the following sentence to the doc comment for `lookup_file_pos()`:

```rust
    ///   ...  If the source_file is empty or the position is located before the
    /// first line, a line number of `0` is returned.
```

**Comment by klensy** [comment]
There also `lookup_file_pos_with_col_display`.

**Comment by richkadel** [comment]
Yes, thanks. The same applies in that function's comment as well. We should update both.

**Comment by klensy** [comment]
`lookup_file_pos_with_col_display` called from `lookup_char_pos` which can assign for `struct Loc` field  `line` value `0` that annotated as
```rust
pub struct Loc {
    /// Information about the original source.
    pub file: Lrc<SourceFile>,
    /// The (1-based) line number.
    pub line: usize,
    /// The (0-based) column offset.
    pub col: CharPos,
    /// The (0-based) column offset when displayed.
    pub col_display: usize,
}
```

**Comment by richkadel** [comment]
True (and again, thanks for recognizing these inconsistencies! They should be fixed.)

But if you look at the implementation of `lookup_char_pos()` [prior to my change](https://github.com/rust-lang/rust/blame/d890e64dff3fd2c28635c60cc9315648dd46c8f7/compiler/rustc_span/src/source_map.rs), if `lookup_line(pos)` returned an error, it also returned `0` for the line number. So my changes were consistent with the prior implementation.

So I guess the `Loc` struct `line` field's doc comment is also incomplete, and should also be updated to reflect the possibility that `line` can be zero.

The alternative is to change all of these line variables and the struct field to an `Option<usize>`, but since this would have a lot of ripple effects, I'm assuming we don't want to do that.

**Comment by klensy** [comment]
Yes, i forgot to check previous revisions, sorry.
I'll try to see how `lookup_line` can return None, perhaps this should emit error instead silently using `0` value.

**Comment by richkadel** [comment]
Just to clarify, `lookup_line()` already returns an `Option` and since it can return `None` for an empty file, that *might* not be an `Err` for that function. I don't know if the downstream calls should return None or Err (versus `0`) given their expected usages.

**Comment by richkadel** [comment]
@klensy - Is there a reason you've labeled this related to `A-code-coverage`? Are you seeing code coverage issues related to empty files or other reasons that the line could be zero?

**Comment by richkadel** [comment]
This is not a bug in code coverage. Code coverage works correctly, based on the current behavior of `SourceMap` and `Span` code. This issue is related to the fact that the `SourceMap` and `Span` code behavior behave in a way that might be confusing to users of that API, and that the API documentation is inconsistent with the actual behavior.

@rustbot label -A-code-coverage

**Comment by hkBst** [comment]
@richkadel, @klensy, is this still relevant, and if so did either of you want to propose a fix?
